src/systems/systems/src/air_conditioning/acs_controller.rs

```diff
@@ -0,0 +1,1808 @@
+        {
+            AcStateManager::InFlight(self.into())
+        } else {
+            AcStateManager::EndTakeOff(self.increase_timer(context))
+        }
+    }
+}
+
+transition!(EndTakeOff, InFlight);
+
+#[derive(Copy, Clone)]
+struct InFlight;
+
+impl AcState<InFlight> {
+    fn step(
+        self: AcState<InFlight>,
+        engines: [&impl EngineCorrectedN1; 2],
+        lgciu: [&impl LgciuWeightOnWheels; 2],
+    ) -> AcStateManager {
+        if engines
+            .iter()
+            .all(|&x| x.corrected_n1() < Ratio::new::<percent>(70.))
+            && lgciu
+                .iter()
+                .all(|&a| a.left_and_right_gear_compressed(true))
+        {
+            AcStateManager::BeginLanding(self.into())
+        } else {
+            AcStateManager::InFlight(self)
+        }
+    }
+}
+
+transition!(InFlight, BeginLanding);
+
+#[derive(Copy, Clone)]
+struct BeginLanding;
+
+impl AcState<BeginLanding> {
+    fn step(
+        self: AcState<BeginLanding>,
+        context: &UpdateContext,
+        engines: [&impl EngineCorrectedN1; 2],
+    ) -> AcStateManager {
+        if (engines
+            .iter()
+            .all(|&x| x.corrected_n1() < Ratio::new::<percent>(70.))
+            && context.indicated_airspeed().get::<knot>() < 70.)
+            || self.timer > Duration::from_secs(35)
+        {
+            AcStateManager::EndLanding(self.into())
+        } else {
+            AcStateManager::BeginLanding(self.increase_timer(context))
+        }
+    }
+}
+
+transition!(BeginLanding, EndLanding);
+
+#[derive(Copy, Clone)]
+struct EndLanding;
+
+impl AcState<EndLanding> {
+    fn step(self: AcState<EndLanding>, context: &UpdateContext) -> AcStateManager {
+        if self.timer > Duration::from_secs(10) {
+            AcStateManager::OnGround(self.into())
+        } else {
+            AcStateManager::EndLanding(self.increase_timer(context))
+        }
+    }
+}
+
+transition!(EndLanding, OnGround);
+
+#[derive(Clone)]
+struct ZoneController {
+    zone_temp_id: VariableIdentifier,
+    zone_duct_temp_id: VariableIdentifier,
+
+    zone_id: &'static str,
+    duct_demand_temperature: ThermodynamicTemperature,
+    zone_selected_temperature: ThermodynamicTemperature,
+    zone_measured_temperature: ThermodynamicTemperature,
+    pid_controller: PidController,
+}
+
+impl ZoneController {
+    const K_ALTITUDE_CORRECTION: f64 = 0.0000375; // deg/feet
+    const UPPER_DUCT_TEMP_LIMIT_LOW: f64 = 323.15; // K
+    const UPPER_DUCT_TEMP_LIMIT_HIGH: f64 = 343.15; // K
+    const LOWER_DUCT_TEMP_LIMIT_LOW: f64 = 275.15; // K
+    const LOWER_DUCT_TEMP_LIMIT_HIGH: f64 = 281.15; // K
+    const KI_DUCT_DEMAND_CABIN: f64 = 0.05;
+    const KI_DUCT_DEMAND_COCKPIT: f64 = 0.04;
+    const KP_DUCT_DEMAND_CABIN: f64 = 3.5;
+    const KP_DUCT_DEMAND_COCKPIT: f64 = 2.;
+
+    fn new(context: &mut InitContext, zone_id: &'static str) -> Self {
+        let pid_controller = if zone_id == "CKPT" {
```

Comment:
**suggestion**: I'm not sure to what extent you'll apply the index-based suggestion I made earlier. Here you could also consider using an enum instead. Something like:

```rust
let pid_controller = match zone_type {
    ZoneType::Cockpit => { ... },
    ZoneType::Cabin(_) => { ... }
}
```

You could then add an implementation for `Display` to the `ZoneType` enum for the `zone_temp_id` and other assignments. See `ElectricalBusType` for an example.

Reply:
As per one of the other comments above. I've now replaced all `Hashmap`s for vectors, const generic arrays and enums. I like it because it makes the code much more "adaptable" to other types of aircraft that might come in the future.

In this particular instance I've implemented your suggestion, but I failed to extract the `Display` implementation from the `systems` module. This is the only bit of code specific to the A320 still living in systems. Do you have any idea on how I can get this out to the `a320_systems` module?

Reply:
Maybe by including the name into the enum? E.g. `ZoneType::Cabin(1, "FWD".to_owned())`. Then in pattern matching most of the time you can just ignore the string (`ZoneType::Cabin(number, _)`), unless you need it.